packages/template-retail-react-app/app/pages/account/order-detail.jsx

```diff
@@ -31,18 +31,87 @@ import CartItemVariantImage from '../../components/item-variant/item-image'
 import CartItemVariantName from '../../components/item-variant/item-name'
 import CartItemVariantAttributes from '../../components/item-variant/item-attributes'
 import CartItemVariantPrice from '../../components/item-variant/item-price'
+import PropTypes from 'prop-types'
+
+const OrderProducts = ({productItems, currency, isLoadingOrder}) => {
+    const ids = productItems.map((item) => item.productId).join(',') ?? ''
+    const {data: {data: products} = {}, isLoading: isLoadingProducts} = useProducts(
+        {
+            parameters: {
+                ids: ids
+            }
+        },
+        {
+            enabled: !isLoadingOrder
+        }
+    )
+
+    const isLoading = isLoadingProducts || isLoadingOrder
+
+    const variants =
+        products &&
+        products.map((product) => {
+            const productItem = productItems.find((item) => item.productId === product.id)
```

Comment:
Searching the array for every product could be expensive. If you build a map you only need to do it once.
```js
const productItemsMap = productIems.reduce((map, item) => ({...map, [item.productId]: item}), {})
```
As a bonus, `ids` (above) can also be changed to use `Object.keys(productItemsMap)`.

Reply:
Good point! We can transform the data from useProducts here.

Reply:
Agree, we go from a loop within a loop O(n²) to linear complexity O(n). I made the changes.